### PR TITLE
Upgrade to SDK 34

### LINKED ISSUE #5770: Bump target SDK to 34
Google has started nudging us to bump our target SDK to API level 34. We have until August 31. After that we won't be able to publish updates unless we bump our target SDK. 

Volunteers are welcome to help us with this. We would specifically need to do the required changes verify our functionality doesn't break as a result of bumping the target SDK.

### PR DESCRIPTION
**Pull request by rohit9625**
**Description (required)**

Fixes #5770 

> **Note:-** Please test these changes and give your feedback.  From this change, the normal gallery picker prompts the user to the photo picker, which is not the desired behavior and it requires major changes. **This PR is testing only for Now. Don't consider it to be production-ready.**

What changes did you make and why?
**1. Add functionality to work with `Partial Access` on API >= 34**
-  This includes adding a new UI component to manage partially accessed photos.
- **Add Jetpack Compose to project**:-Add dependencies to create that new UI component using **Jetpack Compose** in `ComposeView`.
- Perform some refactoring on related classes and methods.
- Add `READ_MEDIA_VISUAL_USER_SELECTED` permission in **Manifest File** and limit some permissions to a specific SDK.

**2. Upgrade AGP to latest and targetSDK/compileSDK to 34**

**3.  Minor refactoring**

- Some changes are not related to this migration but while exploring I found some changes that were needed
- Migrating to `Partial Access` was causing some crashes for specific actions. That's why I need to make some changes to files under the `Custom Selector` package.
- Reduce compiler warnings by adding final keywords to some files(that I saw while exploring)

**4. Replace deprecated circular progress bar**
- Remove the deprecated progress bar from `activity_quiz_result` and `fragment_achievements` and add new **material circular progress bar**

**Tests performed (required)**

Tested ProdDebug on Samsung A14(**API 34**), Realme Narzo 50(**API 33**), Pixel 5 Emulator(**API 29**), and Pixel 6 Emulator(**API 31**).

**Screenshots (for UI changes only)**
>New UI Component

### CODE REVIEW

**Comment by nicolas-raoul** [comment]
I get `Could not find com.dinuscxj:circleprogressbar:1.1.1`.

This repository/branch solves the problem, so would you mind pulling from it? git@github.com:kanahia1/apps-android-commons.git:issue5583
Or just the circleprogressbar parts if you prefer, but pulling the whole branch would be more efficient.

Thanks a lot! :-)

**Comment by rohit9625** [comment]
> I get `Could not find com.dinuscxj:circleprogressbar:1.1.1`.
> 
> This repository/branch solves the problem, so would you mind pulling from it? [git@github.com:kanahia1/apps-android-commons.git:issue5583](https://github.com/kanahia1/apps-android-commons/tree/issue5583) Or just the circleprogressbar parts if you prefer, but pulling the whole branch would be more efficient.
> 
> Thanks a lot! :-)


That branch has so many changes and I looked at the progress bar part. The code was removed and commented out as I did on my local machine. What about I add a new progress bar using Jetpack and push the latest changes or wait for kanahia's PR to get merged?

File: app/src/main/AndroidManifest.xml
Comment (RitikaPahwa4444) [inline app/src/main/AndroidManifest.xml:24]: We use foreground service in our app to ensure contributions get uploaded reliably even when users are using 'Optimised' battery mode. Targeting Android 14 makes it [mandatory to specify a foreground service type](https://developer.android.com/about/versions/14/behavior-changes-14#fgs-types). Could you check this once?

**Review by RitikaPahwa4444 — CHANGES_REQUESTED** [review]
The new UI looks good! :)

Haven't reviewed completely or tested yet, but skimmed through it and had some feedback to share.

File: app/src/main/AndroidManifest.xml
Comment (rohit9625) [reply app/src/main/AndroidManifest.xml:24]: Yes, I checked that but I couldn't find that service in the `Manifest` file, and also don't know what type of service should I assign to it. Only two services are defined in the Manifest file. Can you help me with that?

**Comment by nicolas-raoul** [comment]
> "_What about I add a new progress bar using Jetpack_"

That would be fantastic, thanks a lot. 🙂
Even a simple label x/y would be enough for now.

**Comment by rohit9625** [comment]
> > "_What about I add a new progress bar using Jetpack_"
> 
> That would be fantastic, thanks a lot. 🙂 Even a simple label x/y would be enough for now.

The fragments were built in Java so I had to replace the circular progress bar with the material one(XML Views).

**Comment by rohit9625** [comment]
Hey @nicolas-raoul, please try opening settings from the app. On my device, the app is crashing and maybe, this is the reason that `testSetTotalUploadCount` is failing.

**Comment by nicolas-raoul** [comment]
It seems like I am able to open settings (GSoC branch):

**Comment by nicolas-raoul** [comment]
I just tested this branch, it builds fine. 🙂

The 3 stats (for an empty account) all appear differently, is it expected? If difficult don't worry, priority is low so we can make this a different issue. 👍

**Comment by rohit9625** [comment]
> I just tested this branch, it builds fine. 🙂
> 
> The 3 stats (for an empty account) all appear differently, is it expected? If difficult don't worry, priority is low so we can make this a different issue. 👍
> 
> 

I didn't checked for an empty account. Checking if it is minor bug, then I'll fix it. Can you please try opening the settings. It's crashing and I found the fix but want to ensure if it is happening due to recent changes.

**Comment by nicolas-raoul** [comment]
Ah yes Settings are crashing on this branch.

**Comment by rohit9625** [comment]
> It seems like I am able to open settings

Well, I think that upgrading to SDK 34 has some other changes for `AppCompatActivity` class. The error I encountered:-

> **java.lang.IllegalStateException:  AppCompat has already installed itself into the Window**

That is because **The AppCompatActivity class already takes care of creating and managing the AppCompatDelegate instance for us.**

And we are managing `AppCompatDelegate` in our settings activity. Removing that will solve the bug.

**Comment by rohit9625** [comment]
> I just tested this branch, it builds fine. 🙂
> 
> The 3 stats (for an empty account) all appear differently, is it expected? If difficult don't worry, priority is low so we can make this a different issue. 👍

I don't have an empty account. Please let me know if there's any way that I can test this case. Is there any dummy account or do I need to create a new account?

**Comment by nicolas-raoul** [comment]
Don't worry about empty accounts. Would you mind just posting a screenshot showing how it looks like on your account?

**Comment by rohit9625** [comment]
> Don't worry about empty accounts. Would you mind just posting a screenshot showing how it looks like on your account?

I have added the screenshot in the description of this PR.

**Comment by nicolas-raoul** [comment]
I was able to test with a non-empty account now, looks great! :-) Do not worry about empty accounts.

**Comment by rohit9625** [comment]
I found out the reason behind the crash. I added this dependency when migrating to Compose because the activity having `Composables` was crashing. As far as I remember the reason was related to lifecycle and on the internet I found this dependency to include.
image

But, now everything is working fine after removing this. However, `AppCompat` automatically manages the **`AppCompatDelegate`**. The `SettingsActivity` manages that delegate itself, causing the crash.

File: app/src/main/AndroidManifest.xml
Comment (rohit9625) [reply app/src/main/AndroidManifest.xml:24]: We are using WorkManager to show the notification while uploading contributions. So, we don't need to specify the type of foreground service when using WorkManager.
However, two services are defined in the AndroidManifest file and SDK Upgrade Assistant suggests specifying a type for them. Those services are not foreground services, I guess. What do you think?

Comment (RitikaPahwa4444) [reply app/src/main/AndroidManifest.xml:24]: I haven't gone through the complete documentation yet, but would you mind checking https://developer.android.com/develop/background-work/services/fg-service-types#data-sync and sharing your thoughts?

Comment (rohit9625) [reply app/src/main/AndroidManifest.xml:24]: Thank you @RitikaPahwa4444 for referencing these docs. 
As you said earlier, we are using foreground service for showing the uploading status but I got confused because we didn't have any foreground service defined in the manifest file.
WorkManager automatically utilizes a system-provided foreground service for showing notifications. I am adding the type for this service. Do we have more foreground services except this one?

Comment (RitikaPahwa4444) [reply app/src/main/AndroidManifest.xml:24]: > we didn't have any foreground service defined in the manifest file.

Could you please elaborate? As far as I remember, this existed way before we added it to the upload worker.

**Comment by nicolas-raoul** [comment]
Would you mind rebasing (or pulling) from main? Thanks! :-)

File: app/src/main/AndroidManifest.xml
Comment (rohit9625) [reply app/src/main/AndroidManifest.xml:24]: Which one is a foreground service? Because these are the only two services defined in the `Manifest` file.

     <service
      android:name=".auth.WikiAccountAuthenticatorService"
      android:exported="true"
      android:process=":auth">
      <intent-filter>
        <action android:name="android.accounts.AccountAuthenticator" />
      </intent-filter>
      <meta-data
        android:name="android.accounts.AccountAuthenticator"
        android:resource="@xml/authenticator" />
    </service>

    <service
      android:name="org.acra.sender.SenderService"
      android:exported="false"
      android:process=":acra" />

**Comment by nicolas-raoul** [comment]
Unit tests pass and the app seems to work fine, I just uploaded a few pictures using this branch :-)
Is there any blocker before merging this? The discussion on permissions?

**Comment by rohit9625** [comment]
I also think that migration is complete but I have a confusion about this service defined in the XML:-
https://github.com/commons-app/apps-android-commons/pull/5790#discussion_r1739998508

I am waiting for @RitikaPahwa4444 suggestions on this. 
Do you have any idea about this?

**Comment by RitikaPahwa4444** [comment]
Apologies for missing out on the comment. I've not worked closely with those services and will have to check and confirm😅

**Comment by rohit9625** [comment]
No problem, I am waiting for your review :)

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java
Comment (RitikaPahwa4444) [inline app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:168]: Is this no longer required? If yes, could you please remove it?

File: app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt
Comment (RitikaPahwa4444) [inline app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt:84]: Is this change intended?

**Review by RitikaPahwa4444 — CHANGES_REQUESTED** [review]
Skimmed through the code again, had a few minor suggestions. Need some time to upgrade my Android Studio (the AGP version in this PR is not compatible with mine) but will hopefully test it soon. Thanks for refactoring the code as well 🙂

File: app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt
Comment (rohit9625) [reply app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt:84]: I made this change because there was a typo error in Android Studio. However, I didn't notice that this was about **`lateinit`** property. I'll change it if you want.

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java
Comment (rohit9625) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:168]: **`ACCESS_MEDIA_LOCATION`** is required to read the location information from the pictures. But, I don't know when this permission should be invoked. 
So, I just commented that to prevent invoking it on the App's startup, the app should only ask for permission when needed. I just forgot about this permission, thank you for looking at it :+1: 
Do you know at which place the app tries to read the picture's location?

File: app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt
Comment (RitikaPahwa4444) [reply app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt:84]: Yes, please. 

Tip: Most of the codebases require refactoring after some time and IDEs show up various suggestions. Please feel free to open new issues as and when you find it relevant. This approach would also help in limiting the scope of any PR and discussing more about such suggestions independently as we would be able to break the huge task of refactoring into doable units :)

Comment (rohit9625) [reply app/src/main/java/fr/free/nrw/commons/upload/worker/UploadWorker.kt:84]: Got it, Thanks :)

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java
Comment (RitikaPahwa4444) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:168]: Any time we upload any image, we require this permission. We also ask to access storage and images in the very beginning even though that's required when users use normal/custom selector.

Comment (rohit9625) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:168]: Well in this PR, the storage permission is asked from the user when the user tries to open `custom selector`. However, look at this documentation:-
https://developer.android.com/training/data-storage/shared/media#location-info-photos

Does it mean, we don't need ACCESS_MEDIA_LOCATION permission to read photo's location.

Comment (RitikaPahwa4444) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:168]: Location information probably got stripped off😞
I'm currently testing this PR. I allowed access all from my Gallery and got the no location found pop up. The edit location button shows a green tick, though.

Comment (rohit9625) [reply app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java:168]: I am making a minor change to ask for `ACCESS_MEDIA_LOCATION` permission along with other storage-related permissions. Please test the behavior again. Give me 5 mins only.

**Comment by RitikaPahwa4444** [comment]
Any idea why the app is no longer showing notification for the foreground service or even the ongoing uploads? We currently display two of them, but on this branch I see none.

**Comment by rohit9625** [comment]
> Any idea why the app is no longer showing notification for the foreground service or even the ongoing uploads? We currently display two of them, but on this branch I see none.

No idea :(
I am looking into it. Can you please test the location problem again?

**Comment by rohit9625** [comment]
Both notifications are showing when I am uploading new images. Make sure you have notification permission.

**Comment by RitikaPahwa4444** [comment]
It didn't ask me for notification permission at all. Happened on a fresh installation.

**Comment by rohit9625** [comment]
But, it asked me when I opened that newly implemented screen for managing uploads. Same behavior on the main branch. I think the app is asking for notification permission only when needed or after some time, not sure either.

**Comment by RitikaPahwa4444** [comment]
> I think the app is asking for notification permission only when needed or after some time, not sure either.

I've tried reinstalling the app to observe if that notification permission pops up, but I'm still waiting. It's been over 15 minutes already, I'm uploading pictures but not getting any notification or any dialog requesting access.

**Comment by RitikaPahwa4444** [comment]
Custom selector is now working fine 🎉 The location info is not getting redacted andi can see the Commons logo on already uploaded images now

**Comment by rohit9625** [comment]
> Custom selector is now working fine 🎉 The location info is not getting redacted andi can see the Commons logo on already uploaded images now

What is the procedure to repeat that location reading process? Need to check if that is working fine on the main or not.

**Comment by RitikaPahwa4444** [comment]
Closed and reopened the app and got this crash: 

```
STACK_TRACE=com.bumptech.glide.load.engine.CallbackException: Unexpected exception thrown by non-Glide code
at com.bumptech.glide.load.engine.EngineJob.callCallbackOnResourceReady(EngineJob.java:161)
at com.bumptech.glide.load.engine.EngineJob$CallResourceReady.run(EngineJob.java:428)
at android.os.Handler.handleCallback(Handler.java:1000)
at android.os.Handler.dispatchMessage(Handler.java:104)
at android.os.Looper.loopOnce(Looper.java:242)
at android.os.Looper.loop(Looper.java:362)
at android.app.ActivityThread.main(ActivityThread.java:8393)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:992)
Caused by: java.lang.IllegalStateException: Fragment ExploreMapFragment{dd4f3da} (c3906b03-bb95-4c56-8865-cf6135e5da37) not attached to a context.
at androidx.fragment.app.Fragment.requireContext(Fragment.java:900)
at androidx.fragment.app.Fragment.getResources(Fragment.java:964)
at fr.free.nrw.commons.explore.map.ExploreMapFragment.addMarkerToMap(ExploreMapFragment.java:649)
at fr.free.nrw.commons.explore.map.ExploreMapFragment.addMarkersToMap(ExploreMapFragment.java:636)
at fr.free.nrw.commons.explore.map.ExploreMapPresenter.onNearbyBaseMarkerThumbsReady(ExploreMapPresenter.java:182)
at fr.free.nrw.commons.explore.map.ExploreMapController$1.onResourceReady(ExploreMapController.java:180)
at fr.free.nrw.commons.explore.map.ExploreMapController$1.onResourceReady(ExploreMapController.java:170)
at com.bumptech.glide.request.SingleRequest.onResourceReady(SingleRequest.java:639)
at com.bumptech.glide.request.SingleRequest.onResourceReady(SingleRequest.java:578)
at com.bumptech.glide.load.engine.EngineJob.callCallbackOnResourceReady(EngineJob.java:159)
... 9 more
```
Please don't worry if this is not related to your changes. Let's keep a track of this, though. I'll raise a separate issue if I face it frequently on the main branch as well.

**Comment by rohit9625** [comment]
> > I think the app is asking for notification permission only when needed or after some time, not sure either.
> 
> I've tried reinstalling the app to observe if that notification permission pops up, but I'm still waiting. It's been over 15 minutes already, I'm uploading pictures but not getting any notification or any dialog requesting access. 

Did you try the same on the main. Because even on main branch the app is not requesting the notifications. I tried it myself just few minutes ago.

**Comment by rohit9625** [comment]
> Please don't worry if this is not related to your changes. Let's keep a track of this, though. I'll raise a separate issue if I face it frequently on the main branch as well.

Okay 👍, that'll be nice :)
Even I also noticed a behavior related to pausing uploads. I'll create a new issue for that.

**Comment by RitikaPahwa4444** [comment]
> Because even on main branch the app is not requesting the notification permission

But you still see the notifications, right? 🤔

**Comment by rohit9625** [comment]
> But you still see the notifications, right? 🤔

No, I tried uploading a new picture from the main branch version but no notifications showed up. So, I opened the screen where we can manage uploads and then the app asked me for notification permission. After that, the uploading seemed to restart (Not sure) and both notifications were showing.

**Comment by RitikaPahwa4444** [comment]
I've opened the pending uploads screen several times on this branch but didn't get that permission popup. Testing on main now.

**Comment by rohit9625** [comment]
> I've opened the pending uploads screen several times on this branch but didn't get that permission popup. Testing on main now.

Just to be sure. Did you open that screen just after starting a new upload? Because ongoing uploads are the reason to ask for notification permission, not opening that screen only. However, it seems another issue and out of the scope of this PR. What do you think?

**Comment by RitikaPahwa4444** [comment]
> Did you open that screen just after starting a new upload? 

Yes.

> However, it seems another issue and out of the scope of this PR

Bumping versions does break things. And if notifications for foreground services have disappeared without the user dismissing them, we cannot rest assured that they would work as intended without tesing completely. I'm sorry if you found this unrelated - I'm just testing and sharing what is different on this branch.

**Comment by rohit9625** [comment]
> Custom selector is now working fine 🎉 The location info is not getting redacted and I can see the Commons logo on already uploaded images now

I tried uploading a picture that has a location in its EXIF and the app can read the location info from EXIF of the image after the recent commit. However, I am getting these errors on the logs on this PR and main as well.

image

**Comment by rohit9625** [comment]
> > However, it seems another issue and out of the scope of this PR
> 
> Bumping versions does break things. And if notifications for foreground services have disappeared without the user dismissing them, we cannot rest assured that they would work as intended without testing completely. I'm sorry if you found this unrelated - I'm just testing and sharing what is different on this branch.

You are right, bumping to a new SDK definitely can break other parts. However, it's working on API 34 but the behavior could be different on other versions. What API level you are testing on? 
Thank you once again for your insights :)

**Comment by RitikaPahwa4444** [comment]
I'm testing on Android 14 (API level 34).

**Comment by rohit9625** [comment]
> I'm testing on Android 14 (API level 34).

Now, what am I supposed to do? I am so confused :(

**Comment by rohit9625** [comment]
@RitikaPahwa4444, I can make the app request **`Notification Permission`** at the startup to solve the problem. However, users can decline the permission. Would you happen to have any reference to the current implementation of requesting notification permission?

**Comment by RitikaPahwa4444** [comment]
Hey @rohit9625, I would need some time to check on the notifications issue along with the issue mentioned in [this](https://github.com/commons-app/apps-android-commons/pull/5790#issuecomment-2351513079) comment. Until then, please don't consider them as a blocker. The location issue seems sorted for me at least, but I'll share if I find any errors in my logs. Keeping the notifications issue aside, uploads seem to work fine for me while I'm using the app.

**Review by nicolas-raoul — APPROVED** [review]

**Comment by nicolas-raoul** [comment]
I believe Ritika wrote that there is no blocker, so I will merge this.
I uploaded hundreds of pictures with this branch and it worked fine. Unit tests passing too.
Thanks a lot @rohit9625, this was high priority because it was blocking further releases.

**Pull request merged**
